File: Locker/Number of Closed Islands.cpp
class Solution {
public:
    int closedIsland(vector<vector<int>>& grid) {
        int m = grid.size(), n = grid[0].size();
        for(int i = 0; i < m; i++) {
            for(int j = 0; j < n; j++) {
                if(i == 0 || j == 0 || i == m-1 || j == n-1) {
                    if(grid[i][j] == 0)
                        dfs(grid, i, j, m, n);
                }
            }
        }

        int count = 0;
        for(int i = 1; i < m-1; i++) {
            for(int j = 1; j < n-1; j++) {
                if(grid[i][j] == 0) {
                    count++;
                    dfs(grid, i, j, m, n);
                }
            }
        }
        return count;
    }

    void dfs(vector<vector<int>>& grid, int i, int j, int&m , int& n) {
        if(i < 0 || j < 0 || i >= m || j >= n || grid[i][j] == 1)
            return;
        
        grid[i][j] = 1;
        dfs(grid, i+1, j, m, n);
        dfs(grid, i-1, j, m, n);
        dfs(grid, i, j+1, m, n);
        dfs(grid, i, j-1, m, n);
    }
};

File: Number of Closed Islands.py
class Solution:
    def closedIsland(self, grid: List[List[int]]) -> int:
        m, n = len(grid), len(grid[0])

        def dfs(i, j):
            if i < 0 or j < 0 or i >= m or j >= n or grid[i][j] != 0:
                return
            
            grid[i][j] = 1
            for dx, dy in [(1, 0), (-1, 0), (0, 1), (0, -1)]:
                dfs(i + dx, j + dy)

        for i in range(0, m):
            for j in range(0, n):
                if i == 0 or j == 0 or i == m-1 or j == n-1:
                    dfs(i, j)

        count = 0
        for i in range(1, m-1):
            for j in range(1, n-1):
                if grid[i][j] == 0:
                    count += 1
                    dfs(i, j)
        return count
